packages/core/src/utils/SIWXUtil.ts

```diff
@@ -1,4 +1,256 @@
+import { StorageUtil } from './StorageUtil.js'
+
+export const SIWXUtil = {
+  getSIWX() {
+    return OptionsController.state.siwx
+  },
+  async initializeIfEnabled() {
+    const siwx = OptionsController.state.siwx
+    const address = CoreHelperUtil.getPlainAddress(ChainController.getActiveCaipAddress())
+    const network = ChainController.getActiveCaipNetwork()
+
+    if (!(siwx && address && network)) {
+      return
+    }
+
+    try {
+      const sessions = await siwx.getSessions(network.caipNetworkId, address)
+      if (sessions.length) {
+        return
+      }
+
+      await ModalController.open({
+        view: 'SIWXSignMessage'
+      })
+    } catch (error: unknown) {
+      // eslint-disable-next-line no-console
+      console.error('SIWXUtil:initializeIfEnabled', error)
+
+      EventsController.sendEvent({
+        type: 'track',
+        event: 'SIWX_AUTH_ERROR',
+        properties: this.getSIWXEventProperties()
+      })
+
+      // eslint-disable-next-line no-console
+      await ConnectionController._getClient()?.disconnect().catch(console.error)
+      RouterController.reset('Connect')
+      SnackController.showError('A problem occurred while trying initialize authentication')
+    }
+  },
+  async requestSignMessage() {
+    const siwx = OptionsController.state.siwx
+    const address = CoreHelperUtil.getPlainAddress(ChainController.getActiveCaipAddress())
+    const network = ChainController.getActiveCaipNetwork()
+    const client = ConnectionController._getClient()
+
+    if (!(siwx && address && network && client)) {
+      throw new Error('SIWX is not enabled')
+    }
+
+    try {
+      const siwxMessage = await siwx.createMessage({
+        chainId: network.caipNetworkId,
+        accountAddress: address
+      })
+
+      const message = siwxMessage.toString()
+
+      if (StorageUtil.getConnectedConnector() === 'ID_AUTH') {
+        RouterController.pushTransactionStack({
+          view: null,
+          goBack: false,
+          replace: true
+        })
+      }
+
+      const signature = await client.signMessage(message)
+
+      await siwx.addSession({
+        data: siwxMessage,
+        message,
+        signature: signature as `0x${string}`
+      })
+
+      ModalController.close()
+
+      EventsController.sendEvent({
+        type: 'track',
+        event: 'SIWX_AUTH_SUCCESS',
+        properties: this.getSIWXEventProperties()
+      })
+    } catch (error) {
+      const properties = this.getSIWXEventProperties()
+
+      if (properties.isSmartAccount) {
+        SnackController.showError('This application might not support Smart Accounts')
+      } else {
+        SnackController.showError('Signature declined')
+      }
+
+      EventsController.sendEvent({
+        type: 'track',
+        event: 'SIWX_AUTH_ERROR',
+        properties
+      })
+
+      // eslint-disable-next-line no-console
+      console.error('SWIXUtil:requestSignMessage', error)
+    }
+  },
+  async cancelSignMessage() {
```

Comment:
do you need to update session state?

Reply:
`ConnectionController.disconnect` will call the session revoke, inside AppKit siwx is stateless. Maybe I should bring the code from `ConnectionController.disconnect` into another function of `SIWXUtil`, wdyt?